include/swift/SIL/SILBuilder.h

```diff
@@ -31,27 +31,25 @@ class IntegerLiteralExpr;
 class FloatLiteralExpr;
 class SILGlobalVariable;
 
-class SILBuilder {
-  friend class SILBuilderWithScope;
+/// Manage the state needed for a SIL pass across multiple, independent
+/// SILBuilder invocations.
+class SILBuilderContext {
+  friend class SILBuilder;
 
-  SILFunction *F;
-  SILModule &Mod;
+  using InsertedInstVec = SmallVectorImpl<SILInstruction *>;
+
+  SILModule &Module;
 
   /// Allow the SIL module conventions to be overriden within the builder.
   /// This supports passes that lower SIL to a new stage.
-  SILModuleConventions silConv = SILModuleConventions(Mod);
-
-  /// If this is non-null, the instruction is inserted in the specified
-  /// basic block, at the specified InsertPt.  If null, created instructions
-  /// are not auto-inserted.
-  SILBasicBlock *BB;
-  SILBasicBlock::iterator InsertPt;
-  const SILDebugScope *CurDebugScope = nullptr;
-  Optional<SILLocation> CurDebugLocOverride = None;
```

Comment:
The explicit purpose of the debug location override is to survive changes to the insertion point and debug location. It's used in a case where the builder must emit a batch of instructions with the same location (with exceptions for debug intrinsics), and would be cumbersome to do any other way.

Could you preserve the location override in SILBuilderContext?

Reply:
I'm baffled by this. A batch of instructions with the same debug location should always use the same SILBuilder instance. The point of a separate SILBuilderContext is that some of the state was totally independent of the debug location or insertion point.

Reply:
My point is that a location override is independent of the builder's debug location or insertion point, which fits your description of state managed by SILBuilderContext. In a model where you need a new builder instance to change insertion point, it may not be true that a batch of instructions which should have the same debug location will be emitted by a single builder.

Reply:
We have several broken abstractions built up around SILBuilder. DebugLocOverrideRAII is one of those. Each one of those broken abstractions makes the problem worse in the sense that it makes the responsibility of each type less clear, harder to reason about correctness, and harder to spot bugs. Adding a debug location override to SILBuilderContext would break the abstraction and violate separation of concerns. I need this PR so I can use SILBuilder in utilities as it was originally intended, not so we can have more broken abstractions and make it easier to misuse SILBuilder.

The point is, if you see a SILBuilder being instantiated, you should know what debug location it will use. If a pass is doing something more complicated, it's either going to have it's own wrapper around a SILBuilder, or will need to *explicitly* reuse the builder and *explicitly* modify the debug location or insertion point. In the case of DebugLocOverrideRAII, SILGenBuilder should manage it's own current debug location, it's very easy to do this explicitly without breaking the underlying SILBuilder abstractions.

Reply:
I think limiting usage of the location override to SILGenBuilder would be fine, with the caveat that it would require making getSILDebugLocation virtual. If we can agree that the tradeoff is worthwhile, I have no objections to this patch.

Reply:
I can think of many ways to solve SILGen's getSILDebugLocation issue, but it all depends on how code will be migrated, so we should probably discuss that in a different PR. The mere existence of the override mechanism is direct evidence that the API is severely broken and being misused.